File: src/main/java/gargoyle/netsquares/prop/base/BaseROProperty.java
package gargoyle.netsquares.prop.base;

import java.util.Objects;

public abstract class BaseROProperty<T> implements IROProperty<T> {
    protected BaseROProperty() {
    }

    @Override
    public abstract T get();

    @Override
    public int hashCode() {
        return Objects.hash(get());
    }

    @Override
    public boolean equals(Object o) {
        return this == o || o instanceof IROProperty && Objects.equals(get(), ((IROProperty<?>) o).get());
    }

    @Override
    public String toString() {
        return String.format("%s(%s)", getClass().getName(), get());
    }
}

File: src/main/kotlin/gargoyle/netsquares/prop/base/BaseROProperty.kt
package gargoyle.netsquares.prop.base

abstract class BaseROProperty<T : Any> protected constructor() : IROProperty<T> {
    override fun hashCode(): Int {
        return get().hashCode()
    }

    abstract override fun get(): T
    override fun equals(other: Any?): Boolean {
        return this === other || other is IROProperty<*> && get() == other.get()
    }

    override fun toString(): String {
        return String.format("%s(%s)", javaClass.name, get())
    }
}
